### PR TITLE
Azure - SQL Database View Filter for azure.sqlserver resource

### PR DESCRIPTION
**Pull request by axis7818**
this is the first step in addressing #3030 

* add the SqlDatabaseFilter to azure.sqlserver resource.
  * this filter augments the SQL Server model with a list of SQL Databases belonging to that server.
* add testing and documentation for the filter
  * updated the ARM template for sqlserver to include a SQL database
  * updated the SqlServer test cassettes to work off of the updated ARM template resources.

### CODE REVIEW

**Comment by axis7818** [comment]
Question: since I modified the ARM template for the SQL server, should I have re-recorded the cassettes? or left them the way they were and only generate the cassettes for new tests?

**Comment by logachev** [comment]
The change itself looks good to me.

My concern here is the model where filter is used only to augment resource feels broken.
We have different examples, where we augment resources data in filter & apply generic value filters afterwards (`view-instance` for `vms`, `diagnostic-settings`, `effective route table`).

I feel that instead of us inheriting from `ValueFilter` over and over again Custodian needs some sort of `resource extension` model: those extensions will load more resource details if required for the policy.

I don't think Filters should be allowed to modify existing resources, but at the same time we can't load all information for all resources each time because it can be long-running operation (`Effective route table` takes ~7 seconds per NIC).

@kapilt @stefangordon @aluong what are your thoughts here?

Example I have in mind, thinking about this PR:
```
     policies:
       - name: find-sql-servers-with-premium-database-sku
         resource: azure.sqlserver

         extensions: 
             - type: sql-database-view

         filters:
           - type: value
             key: databases[?sku.tier='Premium']
             value: not-null
```

**Comment by stefangordon** [comment]
There is some existing concept of a ChildResourceQuery in c7n.

See this policy:
https://github.com/cloud-custodian/cloud-custodian/blob/ddf0afec1d618567388f239275ec08b7c5a94ab6/tests/test_vpc.py#L601

and:
https://github.com/cloud-custodian/cloud-custodian/blob/dcfa6cc6c7e1b7d15fdf21ef8949855665f1f4fa/c7n/resources/vpc.py#L1478

and:
https://github.com/cloud-custodian/cloud-custodian/blob/master/c7n/query.py#L119

Maybe this concept helps?  I think some identifiers from the parent are included on the children, probably room to extend that if needed.

File: tools/c7n_azure/c7n_azure/resources/sqlserver.py [hunk 1]
```diff
@@ -23,3 +25,20 @@ class resource_type(ArmResourceManager.resource_type):
...
+    schema = type_schema('sql-database-view')
+
+    def __call__(self, i, *args, **kwargs):
+
+        if 'databases' not in i:
+            client = self.manager.get_client()
+            resource_group = i['resourceGroup']
+            server_name = i['name']
+            dbs = client.databases.list_by_server(resource_group, server_name)
+            i['databases'] = [db.as_dict() for db in dbs]
```
Comment (kapilt) [inline tools/c7n_azure/c7n_azure/resources/sqlserver.py]: we always prefix annotations ue, `c7n:databases` on to resources to disambiguate against any intrinsic attributes.

**Comment by kapilt** [comment]
@logachev re generic expansion, multi attribute and related filters are the ones that have common provisions, multi-attr is generally useful when wanting to match on multiple attributes of the sub/related resource. generally we would want to unique the set of fetches across the resource population to avoid duplicate fetches. the 7 seconds per nic on the route table sounds odd, isn't most of that information normalized/common once your dereferenced to the subnet?

**Comment by stefangordon** [comment]
I think we should do it like this:

https://github.com/cloud-custodian/cloud-custodian/pull/3853

It is then trivial to add a bunch of other similar resource scenarios we have....

**Comment by stefangordon** [comment]
re: 7 seconds per route table

The one you are referencing is "Effective Route Table" - so that is an API that calculates effective routes by merging a bunch of different network resources/attributes.  Just getting a normal route table is quick.

**Comment by axis7818** [comment]
@stefangordon your approach looks good to me

**Pull request closed without merging**